### ISSUE TITLE
pacli proposal show STRING -f usage

### ISSUE THREAD

**Issue by buhtignew** [slimcoin-project/pacli#180]
I've noticed that the `pacli proposal show STRING -f` searches for the string in all proposals of all the decks.
Is it intended?
Would it make sense to provide a filter by the deck as well?
_
I've also seen that in the output there is no hint about what deck that proposal was made for. 
Is there a command to retrieve the deck using the proposals id?
Would it make sense providing a decks id in this command's output?

Update
I've found the command to derive the deck's id from the proposal's id as well as the state of the proposal:
`proposal show db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658 -s | grep -e deck -e state | grep -v donation_states`:
```
    'deck': 'a2459e054ce0f600c90be458915af6bad36a6863a0ce0e33ab76086b514f765a',
    'state': 'abandoned',
```

The state of the proposal can be also retrieved with the `proposal show db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658 -s -p state` command, while the output of the `proposal show db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658 -s -p deck` doesn't contain the deck's info:
```
'Value of parameter deck for proposal '
'db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658:'
<pypeerassets.protocol.Deck object at 0x7fe97e9381d0>
```
 Would it make sense to include these two parameters into the `proposal show STRING -f` output?

**Comment by d5000** [comment]
You're correct. The deck ID is a good addition for the output of the command, added it. The state is already displayed if you use `-a`.

I also added a simpler version of the `-i` flag which also displays the deck ID and some more important parameters. The original information, which is quite technical, is now displayed with `-i` is now retrievable with `-i -a` (i.e. it's the "advanced mode" of this branch of `proposal show`).

If you want to search by deck then the `proposal list` command is more appropiate (the logic of this command parts from the deck and derives all proposals from it). There was no search by string for this mode. I added one with `proposal list DECK -f STRING`. This searches in the proposal's ID or description.

Commit: 5dfacdb. If everything's ok this issue can be closed.

**Comment by buhtignew** [comment]
The `proposal show db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658 -s -p state` command I've mentioned above doesn't work anymore on my side, I'm getting:
```
 File "~/.local/bin/pacli", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "~/.local/lib/python3.12/site-packages/pacli/__main__.py", line 479, in main
    fire.Fire({
  File "~/.local/lib/python3.12/site-packages/fire/core.py", line 143, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/fire/core.py", line 477, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
                                ^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/fire/core.py", line 693, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pacli/dt_classes.py", line 372, in show
    return self.__state(pp_data["states"][0], param=param, complete=advanced, simple=basic, quiet=quiet, search=find, debug=debug)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pacli/dt_classes.py", line 430, in __state
    pprint("Value of parameter {} for proposal {}:".format(param, proposal_id))
                                                                  ^^^^^^^^^^^
NameError: name 'proposal_id' is not defined. Did you mean: 'Proposal'?
```
The same for the `proposal show db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658 -s -p deck` command.

> The original information, which is quite technical, is now displayed with `-i` is now retrievable with `-i -a`

The `proposal show db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658 -i` command is providing the same output of the `proposal show db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658 -i -a` at the moment.
_
> The state is already displayed if you use -a.

Because of the previous section I don't know whether you are referring to the `proposal show db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658 -i -a` or to the `proposal show db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658 -s -a`
command in this case.
However the output of the later one is bit long to be able to find the state of the proposal easily. If you think this info is not that vital it's ok for me at this stage (I don't have enough experience with the code to tell whether we need it or not, TBH).
_
Strangely enough the `proposal show db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658 -f` is outputs:
```
Error: Proposal not found.
```
Even if I reduce it to just one first character (`proposal show d -f`) the output doesn't contain the searched proposal.

**Comment by d5000** [comment]
Oh, that were two oversights when adding the functions yesterday. Fixed.

The `-a` argument cannot be used alone, only together with one of `-f / -s / -i`. Help already showed that but contained a small error which I fixed now.

Regarding the `-f` parameter: I had originally the policy that you had to provide either the complete short ID (with the `-m` flag) or search a string in the description, not in the ID.

I now changed it, so it searches in the "ID string", which is the string composed of the first 16 characters of the ID and the description. This is in line with other commands which search for proposals. Searching by the complete ID should not really make sense as you have the other command variants.

You will also get no result if you search for characters of the ID which aren't in the short ID (first 16 characters). I don't know if it would make sense to add this, I think not. I can implement it though if you think it's important, it's a very short edit.

Commit: 1f19d1f

**Comment by buhtignew** [comment]
> The `-a` argument cannot be used alone, only together with one of `-f`

I'm getting the expected output if I run `proposal show db307777dd1438e0 -f`, while if I run `proposal show db307777dd1438e0 -f -a` the output is `Error: Proposal state not found.`.

> Searching by the complete ID should not really make sense as you have the other command variants.

I agree with you about the complete ID, in fact if the user has a complete ID there is no need to search for that string to get the info about the proposal, but in case the user has lost let say one or more last characters of the complete ID (while he was copy-pasting for instance) and he'd like to search using `-f` to see whether there is a proposal with that part of the sting he has, I think in such a case we should provide him the tool to do so. At least he'd expect us to.
By other hand the same can happen with the first characters of the string. 
So basically, although, it doesn't make sense to search for the complete string we should provide the tool to search for whatever part of the string which I guess in essence means the user should be able to use the command for the complete string as well, although it doesn't make that much sense per se.
_
I still think the state of proposal should be effortlessly retrievable, while at the moment (in case of the proposal `37e7f1556c5e192ac782e3576b5723492daebdb884e1ce0a6d8aa89867194595`, for instance) the only command I was able to get an easily readable output with was `proposal show 37e7f1556c5e192ac782e3576b5723492daebdb884e1ce0a6d8aa89867194595 -s -a | grep state | grep -v donation_states | grep -v dt_states`. 
Am I missing some other option available?

**Comment by d5000** [comment]
I implemented the search by full ID. Still not totally convinced but it may make sense in some cases as you pointed out, and it's a change of 4 words in the code. Commit is b384628.

> proposal show 37e7f1556c5e192ac782e3576b5723492daebdb884e1ce0a6d8aa89867194595 -s -a | grep state | grep -v donation_states | grep -v dt_states

Doesn't the flag combination `-s -p state`  work again after yesterday's update? Here it works.

> proposal show db307777dd1438e0 -f -a

Same with that one ... works here. Are you on the correct update? If yes, what's the result of this and the other command that not works with -d / --debug? (In the case yoy get a "red error", it's always good if you post the result with --debug if available).

**Comment by buhtignew** [comment]
> I implemented the search by full ID.

Thank you very much, on my end it works fine.

> Doesn't the flag combination -s -p state work again after yesterday's update? Here it works.

It works, I don't know why I have overlooked it.

>> proposal show db307777dd1438e0 -f -a

> Same with that one ... works here.

It doesn't work on my end.
The output with the `-d` flag is the following:
```
  File "~/.local/lib/python3.12/site-packages/pacli/dt_utils.py", line 264, in find_proposal_state_by_string
    pstates = get_parser_state(provider, deck, force_continue=True).proposal_states
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pypeerassets/at/dt_misc_utils.py", line 221, in get_parser_state
    unfiltered_cards = list((card for batch in pa.get_card_bundles(provider, deck) for card in batch))
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pypeerassets/at/dt_misc_utils.py", line 221, in <genexpr>
    unfiltered_cards = list((card for batch in pa.get_card_bundles(provider, deck) for card in batch))
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pypeerassets/__main__.py", line 202, in get_card_bundles
    for result in th.map(card_bundle_parser, bundles):
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/_base.py", line 608, in map
    fs = [self.submit(fn, *args) for args in zip(*iterables)]
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pypeerassets/__main__.py", line 193, in <genexpr>
    return (card_bundler(provider, deck, i) for i in raw_txns)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pypeerassets/__main__.py", line 158, in card_bundler
    sender=find_tx_sender(provider, tx),
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pypeerassets/pautils.py", line 55, in find_tx_sender
    txid = vin["txid"]
           ~~~^^^^^^^^
KeyError: 'txid'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "~/.local/bin/pacli", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "~/.local/lib/python3.12/site-packages/pacli/__main__.py", line 479, in main
    fire.Fire({
  File "~/.local/lib/python3.12/site-packages/fire/core.py", line 143, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/fire/core.py", line 477, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
                                ^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/fire/core.py", line 693, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pacli/dt_classes.py", line 364, in show
    pp_data = ei.run_command(self.__get, label_or_id, require_states=(state or find), miniid=miniid, label_priority=(not find), advanced=advanced, debug=debug)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pacli/extended_interface.py", line 34, in run_command
    result = c(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pacli/dt_classes.py", line 277, in __get
    proposal_states = du.find_proposal_state_by_string(str(searchstring), advanced=advanced, shortid=miniid)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pacli/dt_utils.py", line 276, in find_proposal_state_by_string
    raise PacliInputDataError("Proposal state not found.")
pacli.extended_interface.PacliInputDataError: Proposal state not found.
```

**Comment by d5000** [comment]
Normally this error (`KeyError: 'txid'`) occurs when a deck is not initialized. But as it's the "official" dPoD deck, this is unlikely.

However, as this command cycles through all dPoD decks, I believe the problem may be that one of the dPoD decks is not initialized on your side (on my side, all dPoD decks were initialized).

Of course nobody should be forced to initialize all dPoD decks because this could be a spam attack vector. So I added a check where uninitialized decks will be simply skipped. I pre-tested it with a new (and unitialized) dPoD deck and it works, in fact I got the same error than you with the new deck before the change.

This of course means that you only can find proposals on initialized decks, but that's anyway not possible differently (at least in a meaningful way).

Commit: e6c36fe
Can be closed if the error is gone.

**Comment by buhtignew** [comment]
It works now, but the output of the `proposal show db307777dd1438e0 -f -a` is just one line longer than the `proposal show db307777dd1438e0 -f` output and it has the same number of lines of the relative `proposal show db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658 -i` command (the `blockheight` is replaced by `state`).
Compared to the number of lines of the `proposal show db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658 -i -a` and `proposal show db307777dd1438e08b582cd4b5a06e6f632f7c5a2329ed99b07e192da9b08658 -s -a` outputs the `proposal show db307777dd1438e0 -f -a` doesn't seem to be that advanced and taken in consideration how much time it needs to provide its output I wouldn't be that sure people would use it normally. 
Maybe we should consider dropping the `-a` flag to be used with the `-f` flag?

**Comment by d5000** [comment]
The reason why it takes so long is that `-f -a` retrieves the complete state of the proposal, above all on the results of both voting rounds.

The information about the "state" is a quite important value. Just for the `-f`` option I consider it important, because this enables you to search proposals for topics (e.g. "development" or "charity") and see which proposals are active.

So I'm not gonna drop it. I have however made some clarifications in the help, that in the case of `-f` the information `-a` adds is only the "activity" state.

Closing here.
